### PR TITLE
Time sync: Remove ClusterStateCache

### LINKED ISSUE #28930: [1.2] address TSC review comments
### Reproduction steps

See https://github.com/project-chip/connectedhomeip/pull/28418#pullrequestreview-1598871384 @cecille 

### Bug prevalence

Always

### GitHub hash of the SDK that was being used

master

### Platform

core

### Platform Version(s)

_No response_

### Type

Common Cluster Logic

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cecille**
It's handy, but oh so big.

Fixes: #28930

### CODE REVIEW

File: src/app/clusters/time-synchronization-server/time-synchronization-server.h [hunk 1]
```diff
@@ -142,10 +142,11 @@ class TimeSynchronizationServer : public FabricTable::Delegate
     static TimeSynchronizationServer sTimeSyncInstance;
     TimeSyncEventFlag mEventFlag = TimeSyncEventFlag::kNone;
 #if TIME_SYNC_ENABLE_TSC_FEATURE
-    Platform::UniquePtr<app::ClusterStateCache> mAttributeCache;
-    Platform::UniquePtr<app::ReadClient> mReadClient;
     chip::Callback::Callback<chip::OnDeviceConnected> mOnDeviceConnectedCallback;
     chip::Callback::Callback<chip::OnDeviceConnectionFailure> mOnDeviceConnectionFailureCallback;
+    Attributes::UTCTime::TypeInfo::DecodableType mTrustedNodeUtcTime;
+    Attributes::Granularity::TypeInfo::DecodableType mTrustedNodeGranularity;
+    Platform::UniquePtr<app::ReadClient> mReadClient;
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.h:149]: Note: we could have a class that has mTrustedNodeUtcTime, mTrustedNodeGranularity, and ReadClient (by value, not pointer) members, then only need. the memory for the time and granularity while in the middle of the update.  Not sure how much that matters in practice, though.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/time-synchronization-server/time-synchronization-server.h (hunk 1 above)
Comment (cecille) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.h]: done.

File: src/app/chip_data_model.gni [hunk 2]
```diff
@@ -262,8 +266,8 @@ template("chip_data_model") {
           "${_app_root}/clusters/${cluster}/DefaultTimeSyncDelegate.cpp",
           "${_app_root}/clusters/${cluster}/TimeSyncDataProvider.cpp",
         ]
-        cflags +=
-            [ "-DTIME_SYNC_ENABLE_TSC_FEATURE=${time_sync_enable_tsc_feature}" ]
+        defines +=
+            [ "TIME_SYNC_ENABLE_TSC_FEATURE=${time_sync_enable_tsc_feature}" ]
```
Comment (bzbarsky-apple) [inline src/app/chip_data_model.gni:270]: Note that this does not entirely fix https://github.com/project-chip/connectedhomeip/issues/28930.  Specifically, the following are left:

* https://github.com/project-chip/connectedhomeip/pull/28418#discussion_r1307792204
* https://github.com/project-chip/connectedhomeip/pull/28418#discussion_r1307792906
* https://github.com/project-chip/connectedhomeip/pull/28418#discussion_r1307795004
* https://github.com/project-chip/connectedhomeip/pull/28418#discussion_r1307795985
* https://github.com/project-chip/connectedhomeip/pull/28418#discussion_r1307798940

**Review by andy31415 — APPROVED** [review]

File: src/app/chip_data_model.gni (hunk 2 above)
Comment (cecille) [reply src/app/chip_data_model.gni:270]: Comments added in the latest commit to address the ways the callbacks should be handled. The callbacks should be the place where we document how to deal with callbacks, though, not individually on every delegate that uses them. The app:: prefixes have all been removed already AFAICT - the only place I see them here are in the ember functions. Removed a couple from the header, but am I missing something?

File: src/app/clusters/time-synchronization-server/time-synchronization-delegate.h [hunk 3]
```diff
@@ -84,7 +84,11 @@ class Delegate
      * local network defined NTP (DHCPv6 -> DHCP -> DNS-SD sources)
      * If the delegate is unable to support any source, it may return an error immediately. If the delegate is going
      * to attempt to obtain time from any source, it returns CHIP_NO_ERROR and calls the callback on completion.
-     * If the delegate successfully obtains the time, it sets the time using the platform time API (SetClock_RealTime)
+     * If the delegate has a time available at the time of this call, it may call the callback synchronously from within
+     * this function.
+     * If the delegate need to reach out asynchronously to obtain a time, it saves this callback to call asynchronously.
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-delegate.h:89]:
```suggestion
     * If the delegate needs to reach out asynchronously to obtain a time, it saves this callback to call asynchronously.
```

File: src/app/clusters/time-synchronization-server/time-synchronization-delegate.h [hunk 4]
```diff
@@ -84,7 +84,11 @@ class Delegate
...
+     * The delegate should track these callbacks in a list to ensure they can be properly cancelled.
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-delegate.h:90]:
```suggestion
     * The delegate should track these callbacks in a CallbackDeque to ensure they can be properly cancelled.
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/chip_data_model.gni (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/chip_data_model.gni:270]: Thank you for the updates!  Looks good.

**Pull request merged**